Objects/longobject.c

```diff
@@ -3611,24 +3611,29 @@ long_richcompare(PyObject *self, PyObject *other, int op)
     Py_RETURN_RICHCOMPARE(result, 0, op);
 }
 
-static void
-long_dealloc(PyObject *self)
+
+#ifdef Py_DEBUG
+static int _is_python_smallint(PyObject *op)
 {
-    /* This should never get called, but we also don't want to SEGV if
-     * we accidentally decref small Ints out of existence. Instead,
-     * since small Ints are immortal, re-set the reference count.
-     */
-    PyLongObject *pylong = (PyLongObject*)self;
+    PyLongObject *pylong = (PyLongObject*)op;
     if (pylong && _PyLong_IsCompact(pylong)) {
         stwodigits ival = medium_value(pylong);
         if (IS_SMALL_INT(ival)) {
             PyLongObject *small_pylong = (PyLongObject *)get_small_int((sdigit)ival);
             if (pylong == small_pylong) {
-                _Py_SetImmortal(self);
-                return;
+                return 1;
             }
         }
     }
+    return 0;
+}
+#endif
+
+static void
+long_dealloc(PyObject *self)
+{
+    // assert the small ints are not deallocated
+    assert(!_is_python_smallint(self));
```

Comment:
I would just wrap this assertion in a `Py_DEBUG`, then I'll re-approve once buildbots pass.

Reply:
Or they can wrap the first function definition with NDEBUG i think

Reply:
NDEBUG works, thanks for the suggestion.